clients/src/main/java/org/apache/kafka/common/protocol/types/Schema.java

```diff
@@ -64,14 +64,23 @@ public void write(ByteBuffer buffer, Object o) {
     }
 
     /**
-     * Read a struct from the buffer
+     * Read a struct from the buffer. Missing fields at the end of the buffer are replaced with
+     * their default values if such fields are optional; otherwise a {@code SchemaException} is
+     * thrown to signify that mandatory fields are missing.
      */
     @Override
     public Struct read(ByteBuffer buffer) {
         Object[] objects = new Object[fields.length];
         for (int i = 0; i < fields.length; i++) {
             try {
-                objects[i] = fields[i].def.type.read(buffer);
+                if (buffer.hasRemaining()) {
+                    objects[i] = fields[i].def.type.read(buffer);
+                } else if (fields[i].def.hasDefaultValue) {
+                    objects[i] = fields[i].def.defaultValue;
```

Comment:
Should we set the default value here or let it be null and handle the defaults in `Struct#getFieldOrDefault`

Reply:
I agree with you and I'd also like to have only one place where we set defaults. But it seems that the methods that call `Struct#getFieldOrDefault` are not the only way to read a `Struct`'s field. And in this case, this could mean that we set a nullable field that has a default to `null` even if it's default might not be `null`. I think it's more safe to be proactive here and set the default right away, since we know it already in this method.

Reply:
Understood. Thanks for the clarification! Yea, that's a tricky aspect of nullable fields with defaults -- how to determine whether the value is explicitly null or has been defaulted to null :)